src/main/kotlin/org/jitsi/nlj/transform/node/AudioRedHandler.kt

```diff
@@ -0,0 +1,267 @@
+
+        return when (audioPacket) {
+            is RedAudioRtpPacket -> ssrcHandler.transformRed(packetInfo)
+            else -> listOf(ssrcHandler.transformAudio(packetInfo))
+        }
+    }
+
+    override fun trace(f: () -> Unit) = f.invoke()
+
+    override fun getNodeStats(): NodeStatsBlock = super.getNodeStats().apply {
+        addString("red_payload_type", redPayloadType?.toString() ?: "null")
+        addString("audio_level_ext_id", audioLevelExtId?.toString() ?: "null")
+        addString("policy", policy.toString())
+        addString("distance", distance.toString())
+        addBoolean("vad_only", vadOnly)
+        addString("loss_threshold", lossThreshold.toString())
+
+        addNumber("red_packets_decapsulated", stats.redPacketsDecapsulated)
+        addNumber("red_packets_forwarded", stats.redPacketsForwarded)
+        addNumber("audio_packets_encapsulated", stats.audioPacketsEncapsulated)
+        addNumber("audio_packets_forwarded", stats.audioPacketsForwarded)
+        addNumber("lost_packets_recovered", stats.lostPacketsRecovered)
+        addNumber("redundancy_packets_added", stats.redundancyPacketsAdded)
+    }
+
+    /**
+     * Handler for a specific stream (SSRC)
+     */
+    private inner class SsrcRedHandler {
+        val sentAudioCache = RtpPacketCache(20)
+
+        /**
+         * Process an incoming audio packet. It is either forwarded as it is, or encapsulated in RED with previous
+         * packets optionally added as redundancy.
+         */
+        fun transformAudio(packetInfo: PacketInfo): PacketInfo {
+            // Whether we need to add RED encapsulation
+            val redPayloadType = redPayloadType
+            val encapsulate = when (redPayloadType) {
+                null -> false
+                else -> when (policy) {
+                    RedPolicy.NOOP, RedPolicy.STRIP -> false
+                    RedPolicy.PROTECT_ALL -> true
+                    // RedPolicy.PROTECT_DOMINANT -> =isDominant
+                }
+            }
+
+            val audioRtpPacket = packetInfo.packetAs<AudioRtpPacket>()
+
+            sentAudioCache.insert(packetInfo.packetAs())
+
+            if (encapsulate) {
+                val redundancy = mutableListOf<RtpPacket>()
+                if (distance == RedDistance.TWO) {
+                    if (redundancy.maybeAddPacket(applySequenceNumberDelta(audioRtpPacket.sequenceNumber, -2))) {
+                        stats.redundancyPacketAdded()
+                    }
+                }
+                if (redundancy.maybeAddPacket(applySequenceNumberDelta(audioRtpPacket.sequenceNumber, -1))) {
+                    stats.redundancyPacketAdded()
+                }
+
+                val redPacket = RedAudioRtpPacket.builder.build(redPayloadType!!, audioRtpPacket, redundancy)
+                packetInfo.packet = redPacket
+
+                redundancy.forEach { BufferPool.returnBuffer(it.buffer) }
```

Comment:
We do this because `sentAudioCache#get` gives us a copy?  Do we actually modify the buffers before building the RED packet?  If not, could we have a non-copying `get` for packets from a cache since they won't be modified?

Reply:
This particular `reuturnBuffer` is needed because `removeRedAndGetRedundancyPackets` allocates new buffers.

However, I looked at the `cache.get` usage, and it does make a copy (that I currently fail to return) and I think can be avoided the way you suggest!